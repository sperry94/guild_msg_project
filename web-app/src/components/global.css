* {
  scrollbar-color: #777 #212121;
}

*::-webkit-scrollbar-track {
  background-color: #212121;
}

*::-webkit-scrollbar-thumb {
  background-color: #777;
}

html {
  width: 100vw;
  height: 100vh;
}

body {
  width: 100%;
  height: 100%;
}

html,
body {
  background-color: #303030;
}

html,
body,
main {
  overflow: hidden;
}
