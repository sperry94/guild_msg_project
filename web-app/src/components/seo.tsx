import React from "react"
import Helmet from "react-helmet"
import { useStaticQuery, graphql } from "gatsby"

export type SEOProps = {
  description?: string
  lang?: string
  meta?: React.DetailedHTMLProps<
    React.MetaHTMLAttributes<HTMLMetaElement>,
    HTMLMetaElement
  >[]
  title?: string
}

export type SEOGraphqlQuery = {
  site: {
    siteMetadata: {
      title: string
      description: string
      author: string
    }
  }
}

function SEO(props: SEOProps) {
  const { title, lang = `en`, meta = [], description = `` } = props

  const { site } = useStaticQuery<SEOGraphqlQuery>(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            author
          }
        }
      }
    `
  )

  const metaDescription = description || site.siteMetadata.description

  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={([
        {
          name: `description`,
          content: metaDescription,
        },
        {
          property: `og:title`,
          content: title,
        },
        {
          property: `og:description`,
          content: metaDescription,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: site.siteMetadata.author,
        },
        {
          name: `twitter:title`,
          content: title,
        },
        {
          name: `twitter:description`,
          content: metaDescription,
        },
      ] as React.DetailedHTMLProps<
        React.MetaHTMLAttributes<HTMLMetaElement>,
        HTMLMetaElement
      >[]).concat(meta)}
    />
  )
}

export default SEO
