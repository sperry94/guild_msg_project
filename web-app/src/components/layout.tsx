import React from "react"
import Toolbar from "./toolbar"
import { makeStyles } from "@material-ui/core/styles"

import "./global.css"

const useStyles = makeStyles(theme => ({
  toolbarOffset: {
    ...theme.mixins.toolbar,
  },
  mainInner: {
    backgroundColor: theme.palette.background.default,
    color: theme.palette.text.primary,
  },
}))

export type LayoutProps = {
  classes?: {
    main?: string
  }
  children: React.ReactNode | React.ReactNodeArray
}

const Layout = ({ classes: propClasses = {}, children }: LayoutProps) => {
  const innerClasses = useStyles()

  const classes = { ...propClasses, ...innerClasses }

  return (
    <div className={classes.mainInner}>
      <Toolbar />
      <div className={classes.toolbarOffset} />
      <main className={classes.main}>{children}</main>
    </div>
  )
}

export default Layout
