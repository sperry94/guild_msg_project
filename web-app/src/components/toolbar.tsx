import React from "react"
import { Link } from "gatsby"
import { makeStyles } from "@material-ui/core/styles"
import {
  AppBar,
  Button,
  Link as MuiLink,
  Toolbar as MuiToolbar,
  Typography,
} from "@material-ui/core"
import { signIn, signOut, useAuthContext } from "../services/auth"

const useStyles = makeStyles({
  appBar: {
    zIndex: 3,
    height: "64px",
  },
  header: {
    display: "flex",
    height: "64px",
  },
  nav: {
    flexGrow: 1,
    display: "flex",
    justifyContent: "space-between",
    alignItems: "center",
    paddingRight: "24px",
  },
})

const Toolbar = () => {
  const auth = useAuthContext()

  const classes = useStyles()

  return (
    <AppBar color="default" className={classes.appBar}>
      <MuiToolbar className={classes.header}>
        <nav className={classes.nav}>
          <MuiLink component={Link} to="/" color="inherit" underline="none">
            <Typography component="h1" variant="h6" color="inherit">
              Guild Chat
            </Typography>
          </MuiLink>
          {auth && (
            <MuiLink
              component={Link}
              to="/messages"
              color="inherit"
              underline="none"
            >
              <Typography variant="body1" color="inherit">
                My Messages
              </Typography>
            </MuiLink>
          )}
        </nav>
        <div></div>
        {!auth && <Button onClick={() => signIn()}>Log In</Button>}
        {auth && <Button onClick={() => signOut()}>Log Out</Button>}
      </MuiToolbar>
    </AppBar>
  )
}

export default Toolbar
