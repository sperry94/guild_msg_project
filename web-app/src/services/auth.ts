import { createContext, useContext, useState, useEffect } from "react"
import { auth, User } from "firebase/app"
import { firebaseAuth } from "./firebase"

export const AuthContext = createContext<User | null | undefined>(undefined)

export const useAuthContext = () => useContext(AuthContext)

export const useAuthUpdate = () => {
  const [auth, setAuth] = useState<User | null | undefined>(undefined)

  useEffect(() => {
    return firebaseAuth().onAuthStateChanged(setAuth, err => {
      console.error("An error occurred checking auth state", err)
      setAuth(null)
    })
  }, [])

  return auth
}

let innerGoogleAuthProvider: auth.GoogleAuthProvider | null = null

const googleAuthProvider = () =>
  innerGoogleAuthProvider ||
  (innerGoogleAuthProvider = new auth.GoogleAuthProvider())

export const signIn = () =>
  firebaseAuth()
    .signInWithRedirect(googleAuthProvider())
    .catch(err => {
      console.error("An error occurred logging in", err)
    })

export const signOut = () =>
  firebaseAuth()
    .signOut()
    .catch(err => {
      console.error("An error occurred logging out", err)
    })
