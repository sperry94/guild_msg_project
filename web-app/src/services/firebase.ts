import * as firebase from "firebase/app"

import "firebase/auth"

const firebaseConfig = {
  apiKey: "AIzaSyAFkLfldso4y5CJGj8BNCoAv9r7Ukacrow",
  authDomain: "guild-msg-project.firebaseapp.com",
  databaseURL: "https://guild-msg-project.firebaseio.com",
  projectId: "guild-msg-project",
  storageBucket: "guild-msg-project.appspot.com",
  messagingSenderId: "88742455218",
  appId: "1:88742455218:web:602d0164baff3181bd66ab",
  measurementId: "G-WCV6QT9SZ5",
}

let innerFirebaseApp: firebase.app.App | undefined

export const firebaseApp = () =>
  innerFirebaseApp ||
  (innerFirebaseApp = firebase.initializeApp(firebaseConfig))

let innerFirebaseAuth: firebase.auth.Auth | undefined

export const firebaseAuth = () =>
  innerFirebaseAuth || (innerFirebaseAuth = firebaseApp().auth())
