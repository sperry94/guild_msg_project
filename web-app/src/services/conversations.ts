import { useState, useEffect } from "react"
import { firestore, firebaseFunctions } from "./firebase"
import { useAuthContext } from "./auth"

export type User = {
  displayName: string
  email: string
  photoUrl: string
  uid: string
}

type ServerConversation = {
  id: string
  users: string[]
}

export type Conversation = {
  id: string
  users: User[]
}

const cachedUserData: { [k: string]: User } = {}

export const useConversationList = (
  errCb: (msg: string) => void = () => undefined
) => {
  const [conversations, setConversations] = useState<Conversation[] | null>(
    null
  )

  const auth = useAuthContext()

  useEffect(() => {
    if (!auth) {
      setConversations([])
      return
    }

    return firestore()
      .collection("conversations")
      .where("users", "array-contains", auth.uid)
      .onSnapshot(
        async snap => {
          const conversations = snap.docs.map(
            doc =>
              ({
                id: doc.id,
                ...doc.data(),
              } as ServerConversation)
          )

          const allUsers = new Set(
            conversations.flatMap(c => c.users as string[])
          )

          allUsers.delete(auth.uid)

          const userData = await Promise.all(
            Array.from(allUsers)
              .filter(u => cachedUserData[u] === undefined)
              .map(u =>
                firestore()
                  .collection("users")
                  .doc(u)
                  .get()
              )
          )

          for (const u of userData) {
            const uData = u.data() as User
            cachedUserData[uData.uid] = uData
          }

          const convertedConversations: Conversation[] = conversations.map(
            c => ({
              ...c,
              users: c.users
                .filter(u => u !== auth.uid)
                .map(u => cachedUserData[u]),
            })
          )

          setConversations(convertedConversations)
        },
        err => {
          console.error(
            "An error occurred reading the conversations for the current user",
            err
          )
          errCb(
            err?.message ||
              "An error occurred reading the conversations for the current user"
          )
          setConversations([])
        }
      )
  }, [auth, errCb])

  return conversations
}

export const addConversation = async (email: string): Promise<string> => {
  if (!email) {
    throw new Error("An email must be provided to add a new conversation!")
  }

  const res = await firebaseFunctions().httpsCallable("addConversation")({
    email,
  })

  return res.data.docId as string
}
