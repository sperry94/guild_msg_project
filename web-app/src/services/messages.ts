import { useState, useEffect } from "react"
import { firestore as innerFirestore } from "firebase/app"
import { firestore, firebaseAuth } from "./firebase"

export type Message = {
  time: innerFirestore.Timestamp
  senderId: string
  message: string
}

export const useMessages = (conversationId?: string) => {
  const [messages, setMessages] = useState<Message[] | null>(null)

  useEffect(() => {
    if (!conversationId) {
      setMessages([])
      return
    }

    return firestore()
      .collection("conversations")
      .doc(conversationId)
      .collection("messages")
      .orderBy("time")
      .limitToLast(50)
      .onSnapshot(
        snap => {
          setMessages(snap.docs.map(doc => doc.data() as Message))
        },
        err => {
          console.error("An error occurred reading the conversation data", err)
          setMessages([])
        }
      )
  }, [conversationId])

  return messages
}

export const sendMessage = async (conversationId: string, message: string) => {
  if (!conversationId || !message) {
    throw new Error(
      "A conversation ID and a message must be passed in order to save."
    )
  }

  const senderId = firebaseAuth().currentUser?.uid

  if (!senderId) {
    throw new Error("A user must be logged-in in order to save.")
  }

  await firestore()
    .collection("conversations")
    .doc(conversationId)
    .collection("messages")
    .add({
      time: innerFirestore.Timestamp.now(),
      message,
      senderId,
    })
}
