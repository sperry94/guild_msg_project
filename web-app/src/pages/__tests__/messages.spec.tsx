import React from "react"
import { useStaticQuery } from "gatsby"
import renderer from "react-test-renderer"
import { SEOGraphqlQuery } from "../../components/seo"
import { useAuthContext } from "../../services/auth"
import Messages from "../messages"

jest.mock("../../services/auth")
jest.mock("../../services/firebase")
jest.mock("../../services/conversations")
jest.mock("../../services/messages")

const mocked_useAuthContext = useAuthContext as jest.Mock<any>
const mocked_useStaticQuery = useStaticQuery as jest.Mock<SEOGraphqlQuery>

describe("MessagesViewer", () => {
  beforeEach(() => {
    mocked_useStaticQuery.mockReturnValueOnce({
      site: {
        siteMetadata: {
          title: "My title",
          description: "My description",
          author: "Seth Perry",
        },
      },
    })
  })

  it(`renders with auth`, () => {
    mocked_useAuthContext.mockReturnValueOnce({})
    const tree = renderer.create(<Messages />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders without auth`, () => {
    mocked_useAuthContext.mockReturnValueOnce(null)
    const tree = renderer.create(<Messages />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
