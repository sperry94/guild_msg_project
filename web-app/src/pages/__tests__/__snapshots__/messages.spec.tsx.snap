// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesViewer renders with auth 1`] = `
<div
  className="makeStyles-mainInner-4"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed makeStyles-appBar-5 MuiAppBar-colorDefault mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-header-6 MuiToolbar-gutters"
    >
      <nav
        className="makeStyles-nav-7"
      >
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
          href="/"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <h1
            className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
          >
            Guild Chat
          </h1>
        </a>
      </nav>
      <div />
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Log In
        </span>
      </button>
    </div>
  </header>
  <main
    className=""
  >
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div
        className="makeStyles-content-117"
      >
        <div
          className="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-120"
        >
          <div
            className="MuiPaper-root MuiDrawer-paper makeStyles-drawerPaper-121 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
            style={
              Object {
                "visibility": undefined,
              }
            }
          >
            <div
              className="makeStyles-toolbarOffset-119"
            />
            <div
              className="MuiCircularProgress-root makeStyles-spinner-138 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
            >
              <svg
                className="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx={44}
                  cy={44}
                  fill="none"
                  r={20.2}
                  strokeWidth={3.6}
                  style={Object {}}
                />
              </svg>
            </div>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-addNewBtn-122"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
                <p
                  className="MuiTypography-root MuiTypography-body1"
                >
                  New Conversation
                </p>
              </span>
            </button>
          </div>
        </div>
        <div
          className="makeStyles-activeConversation-118 makeStyles-content-179"
        >
          <div
            className="MuiPaper-root makeStyles-content-184 MuiPaper-elevation5"
          >
            <form
              action="#"
              className="makeStyles-form-185"
              onReset={[Function]}
              onSubmit={[Function]}
            >
              <div
                className="MuiFormControl-root MuiTextField-root makeStyles-input-186"
              >
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    aria-label="Message"
                    autoFocus={true}
                    className="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                    disabled={true}
                    name="message"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-228 MuiOutlinedInput-notchedOutline"
                    style={
                      Object {
                        "paddingLeft": 8,
                      }
                    }
                  >
                    <legend
                      className="PrivateNotchedOutline-legend-229"
                      style={
                        Object {
                          "width": 0.01,
                        }
                      }
                    >
                      <span
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="submit"
              >
                <span
                  className="MuiButton-label"
                >
                  Send
                </span>
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`MessagesViewer renders without auth 1`] = `
<div
  className="makeStyles-mainInner-4"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed makeStyles-appBar-5 MuiAppBar-colorDefault mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-header-6 MuiToolbar-gutters"
    >
      <nav
        className="makeStyles-nav-7"
      >
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
          href="/"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <h1
            className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
          >
            Guild Chat
          </h1>
        </a>
      </nav>
      <div />
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Log In
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
  <main
    className="makeStyles-unAuthenticatedMain-2"
  >
    <p
      className="MuiTypography-root MuiTypography-h5"
    >
      You must be logged in to view your messages.
    </p>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiTypography-root MuiTypography-h6"
        >
          Sign in
        </span>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </main>
</div>
`;
